numpy/_core/include/numpy/ndarraytypes.h

```diff
@@ -1300,8 +1300,10 @@ typedef struct {
          */
 #if defined(NPY_INTERNAL_BUILD) && NPY_INTERNAL_BUILD
```

Comment:
On the other hand: let's just add `|| defined(__cplusplus)` here?

That way, we only lie in C++ at least, while in C the old syntax is nice, because it removes any promise of struct size.

Reply:
> let's just add || defined(__cplusplus)

Done, added an explanatory comment too.

Reply:
Hmm, this breaks the ability of SciPy to retain runtime compatibility with NumPy `1.22.4` per gh-26060.

Is the decision here between retaining backward compatibility in SciPy and just someone who noticed a warning with pedantic compiler flags? Or do you have a genuine concern that this will actually cause problems in C++ beyond the warning?